src/libopensc/laser.c

```diff
@@ -0,0 +1,1630 @@
+/*
+ * pkcs15-laser.c: Support for JaCarta PKI applet
+ *
+ * Copyright (C) 2025  Andrey Khodunov <a.khodunov@aladdin.ru>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#ifdef ENABLE_OPENSSL /* empty file without openssl */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <openssl/sha.h>
+#include <openssl/rand.h>
+
+#include "asn1.h"
+#include "cardctl.h"
+#include "common/compat_strlcpy.h"
+#include "internal.h"
+#include "pkcs11/pkcs11.h"
+#include "pkcs15.h"
+#include "laser.h"
+
+#define LOG_ERROR_RET(ctx, r, text) \
+	do { \
+		int _ret = (r); \
+		sc_do_log_color(ctx, SC_LOG_DEBUG_NORMAL, FILENAME, __LINE__, __FUNCTION__, SC_COLOR_FG_RED, \
+				"%s: %d (%s)\n", (text), (_ret), sc_strerror(_ret)); \
+		return (r); \
+	} while (0)
+#define LOG_ERROR_GOTO(ctx, r, text) \
+	do { \
+		int _ret = (r); \
+		sc_do_log_color(ctx, SC_LOG_DEBUG_NORMAL, FILENAME, __LINE__, __FUNCTION__, SC_COLOR_FG_RED, \
+				"%s: %d (%s)\n", (text), (_ret), sc_strerror(_ret)); \
+		goto err; \
+	} while (0)
```

Comment:
Please, move these to the log.h as they sound like not related to this card driver only, but could be useful in other drivers.